src/Symfony/Component/Cache/Marshaller/DefaultMarshaller.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Cache\Marshaller;
+
+/**
+ * Serializes/unserializes values using igbinary_serialize() if available, serialize() otherwise.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class DefaultMarshaller implements MarshallerInterface
```

Comment:
Maybe this has been asked, but why do we have a `DefaultMarshaller` with some internal logic in all methods to differentiate between PHP and IgBinary ... instead of defining a PhpMarshaller and IgBinaryMarshaller (moreover given that we define a MarshallerInterface).

Reply:
Because there is only one implementation of MarshallerInterface: DefaultMarshaller, which auto-adaptatively uses serialize() or igbinary_serialize().

Reply:
But that looks to me like creating a `DefaultMailer` which makes `if($use_smtp) { ... } elseif ($use_gmail) { ...}` inside its methods. In other parts of the framework we define 1 interface and multiple different implementation classes. Here it looks like we have multiple implementations inside a single class. It looks confusing to me, sorry.

Reply:
@javiereguiluz it's not the same. Here, by dealing with both serialization methods (and there are none others for PHP objects), we provide an important feature: auto-adaptative unserialization, providing safe forward and backward compatibility at the data level. This is not possible without having something thatn know about both formats.

Reply:
IgBinaryMarshaller could take PhpMarshaller as a fallback marshaller, no?

Reply:
> IgBinaryMarshaller could take PhpMarshaller as a fallback marshaller

We could build that, but each marshaller would be individually useless, so that'd be a wrong dichotomy to me.

Reply:
That's true ... but maybe we're trying to do too much here? The transparent failover between PHP and IgBinary is nice ... but is that a common enough scenario? Do people enable/disable the igbinary extension "randomly"? Or there are other scenarios where this is useful and I'm missing them? Thanks!

Reply:
@javiereguiluz the day you want to have smaller payload, you'll want to enable igbinary. That day, we can tell you: the data migration is your duty, happy debugging and data migration. Or we could tell: add the extension and the code is ready to deal with that transparently. I'm not sure what's on the other of the balance that would prevent us from doing so.

Reply:
Why would it be individually useless? You could use the PhpMarshaller on its own without ever bothering with igbinary. Or you could use the IgBinaryMarshaller with a strong requirement of never falling back to PHP serialization. I think it gives us the best of both worlds.

Reply:
@teohhanhui what's your describing is a less capable feature (no autoadaptation) with a more complex design.

Reply:
Or you could use the IgBinaryMarshaller but pass in the PhpMarshaller as a fallback marshaller. That should provide the auto-adaptation.

Reply:
That's over-engineering. A combined marshaller is just more capable than each individually, so there is zero reason to create individual ones.

Reply:
It's not over-engineering when it allows other marshallers to easily fallback to PHP serialization without having to implement it on their own.

Reply:
> It's not over-engineering when it allows other marshallers to easily fallback to PHP serialization without having to implement it on their own.

that's correct! and I take that "when" to the letter: when? => never. There are only two php serializers that can reasonably be used here and I don't see any other coming for the foreseeable future. Planning for another eventuality is why I call that over-engineering :) (nothing personal of course).

Reply:
Thanks to the interface, the serialized string itself can still be proceed, e.g compressed and/or encrypted.
One could also build a serializer that would be dedicated to and optimized for very specific data structures (eg only numbers, etc.), etc.

Reply:
Hmm... But actually it's not true that igbinary is the only serialization format that might benefit from a fallback to PHP serialization. There are / were other contenders like msgpack (MessagePack) and Protobuf, both of which have PHP extensions.

Reply:
Nice, another reason to have the interface. Let's keep the topic: should we split the default marshaller in two. I think we would do a misfeature by doing so: things would be more complex to understand, more complex to wire, less capable, and less performant while in a critical code path. Would the individual envisioned marshaller empower our users in comparison to these downsides? Not at all in my opinion.

Reply:
I think you're right, and those other serializer formats are not PHP-specific, so it wouldn't make sense for them to fallback to PHP serialization, actually. :laughing: